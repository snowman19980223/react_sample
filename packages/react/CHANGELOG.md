# @primer/react

## 36.8.0

### Minor Changes

- [#4214](https://github.com/primer/react/pull/4214) [`3dd498e6a7c28a1ec39b56912f76579f6674e022`](https://github.com/primer/react/commit/3dd498e6a7c28a1ec39b56912f76579f6674e022) Thanks [@langermank](https://github.com/langermank)! - Dialog component: add new `position` prop to support fullscreen and bottom sheets on narrow screens

- [#4236](https://github.com/primer/react/pull/4236) [`14a2776f18cf8f4a360b1887db64f8d5e904bc9d`](https://github.com/primer/react/commit/14a2776f18cf8f4a360b1887db64f8d5e904bc9d) Thanks [@joshblack](https://github.com/joshblack)! - Refactor project to avoid circular dependencies

### Patch Changes

- [#4199](https://github.com/primer/react/pull/4199) [`b6e58079a1517d2f1d31098b4e484b94625deabd`](https://github.com/primer/react/commit/b6e58079a1517d2f1d31098b4e484b94625deabd) Thanks [@siddharthkp](https://github.com/siddharthkp)! - experimental/SelectPanel: Improve keyboard navigation from search input

- [#4200](https://github.com/primer/react/pull/4200) [`dc988141216b91561a0ea0943b2cbff03be95029`](https://github.com/primer/react/commit/dc988141216b91561a0ea0943b2cbff03be95029) Thanks [@broccolinisoup](https://github.com/broccolinisoup)! - Tooltip v2: Allow external id to be passed down in tooltip so that the trigger can be labelled/described by the that id

- [#4246](https://github.com/primer/react/pull/4246) [`2aced1c61f26513ef2f7a5a4d05485c6fe7ec1ff`](https://github.com/primer/react/commit/2aced1c61f26513ef2f7a5a4d05485c6fe7ec1ff) Thanks [@siddharthkp](https://github.com/siddharthkp)! - Dialog v1: Fix font-family for title

- [#4265](https://github.com/primer/react/pull/4265) [`19c716f7f1ec44f1f1ee4a135b828208f2be058a`](https://github.com/primer/react/commit/19c716f7f1ec44f1f1ee4a135b828208f2be058a) Thanks [@siddharthkp](https://github.com/siddharthkp)! - SelectPanel2: Submit panel when an item is selected with `Enter` key

- [#4166](https://github.com/primer/react/pull/4166) [`c66b808cbefdbdd7482be752e5d311d2db4d23d8`](https://github.com/primer/react/commit/c66b808cbefdbdd7482be752e5d311d2db4d23d8) Thanks [@pksjce](https://github.com/pksjce)! - Fix focus styles in ActionList Item to support more than focus-visible

- [#4259](https://github.com/primer/react/pull/4259) [`c9fbef6438131cf358e2eaff3d866cf4398e622c`](https://github.com/primer/react/commit/c9fbef6438131cf358e2eaff3d866cf4398e622c) Thanks [@siddharthkp](https://github.com/siddharthkp)! - ActionList: Prevent scroll when an item is selected with `Space`

- [#4245](https://github.com/primer/react/pull/4245) [`d79a419116ea9e669c1ce598144a4b61ae15c52d`](https://github.com/primer/react/commit/d79a419116ea9e669c1ce598144a4b61ae15c52d) Thanks [@siddharthkp](https://github.com/siddharthkp)! - experimental/SelectPanel: Automatically adjust height based on contents

- [#4204](https://github.com/primer/react/pull/4204) [`da4469d9d958733e1b1dd32d6cd1a1bd0661ff91`](https://github.com/primer/react/commit/da4469d9d958733e1b1dd32d6cd1a1bd0661ff91) Thanks [@dylanatsmith](https://github.com/dylanatsmith)! - Increase Tooltip border-radius to match Primer View Components

- [#4205](https://github.com/primer/react/pull/4205) [`ff354604f70fc900cb79fe4c58eaee541a813d9f`](https://github.com/primer/react/commit/ff354604f70fc900cb79fe4c58eaee541a813d9f) Thanks [@pablonete](https://github.com/pablonete)! - Bugfix: Dialog.Header ignored the sx provided
