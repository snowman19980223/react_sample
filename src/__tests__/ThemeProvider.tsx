import {render, screen, waitFor} from '@testing-library/react'
import React from 'react'
import {ThemeProvider, Text, useTheme} from '..'
import 'jest-styled-components'

// window.matchMedia() is not implmented by JSDOM so we have to create a mock:
// https://jestjs.io/docs/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
})

it('respects theme prop', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    space: ['0', '0.25rem']
  }

  render(
    <ThemeProvider theme={theme}>
      <Text color="text" mb={1}>
        Hello
      </Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#f00')
  expect(screen.getByText('Hello')).toHaveStyleRule('margin-bottom', '0.25rem')
})

it('has default theme', () => {
  render(
    <ThemeProvider>
      <Text color="text.primary" mb={1}>
        Hello
      </Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toMatchSnapshot()
})

it('defaults to light color scheme', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme}>
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#00f')
})

it('defaults to dark color scheme in night mode', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="night">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#0f0')
})

it('respects nightScheme prop', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="night" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#ff0')
})

it('respects dayScheme prop', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="day" dayScheme="dark" nightScheme="dark_dimmed">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#0f0')
})

it('works in auto mode', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="auto">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#00f')
})

it('works in auto mode (dark)', () => {
  const matchMediaSpy = jest.spyOn(window, 'matchMedia').mockImplementation(query => ({
    matches: true, // enable dark mode
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))

  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="auto">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#0f0')

  matchMediaSpy.mockRestore()
})

it('works in auto mode', () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  render(
    <ThemeProvider theme={theme} colorMode="auto">
      <Text color="text">Hello</Text>
    </ThemeProvider>
  )

  expect(screen.getByText('Hello')).toHaveStyleRule('color', '#00f')
})

it('updates when colorMode prop changes', async () => {
  const theme = {
    colors: {
      text: '#f00'
    },
    colorSchemes: {
      light: {
        colors: {
          text: '#00f'
        }
      },
      dark: {
        colors: {
          text: '#0f0'
        }
      },
      dark_dimmed: {
        colors: {
          text: '#ff0'
        }
      }
    }
  }

  function App() {
    const [colorMode, setColorMode] = React.useState<'day' | 'night'>('day')
    return (
      <ThemeProvider theme={theme} colorMode={colorMode}>
        <Text color="text">{colorMode}</Text>
        <button onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>Toggle</button>
      </ThemeProvider>
    )
  }

  render(<App />)

  // starts in day mode (light scheme)
  expect(screen.getByText('day')).toHaveStyleRule('color', '#00f')

  screen.getByRole('button').click()

  await waitFor(() =>
    // clicking the toggle button enables night mode (dark scheme)
    expect(screen.getByText('night')).toHaveStyleRule('color', '#0f0')
  )
})

describe('setColorMode', () => {
  it('changes the color mode', () => {
    const theme = {
      colors: {
        text: '#f00'
      },
      colorSchemes: {
        light: {
          colors: {
            text: '#00f'
          }
        },
        dark: {
          colors: {
            text: '#0f0'
          }
        },
        dark_dimmed: {
          colors: {
            text: '#ff0'
          }
        }
      }
    }

    function ToggleMode() {
      const {colorMode, setColorMode} = useTheme()
      return <button onClick={() => setColorMode(colorMode === 'day' ? 'night' : 'day')}>Toggle</button>
    }

    render(
      <ThemeProvider theme={theme} colorMode="day">
        <Text color="text">Hello</Text>
        <ToggleMode />
      </ThemeProvider>
    )

    // starts in day mode (light scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', '#00f')

    screen.getByRole('button').click()

    // clicking the toggle button enables night mode (dark scheme)
    expect(screen.getByText('Hello')).toHaveStyleRule('color', '#0f0')
  })
})
