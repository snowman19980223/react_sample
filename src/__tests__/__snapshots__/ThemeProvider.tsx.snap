// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has default theme 1`] = `
.c0 {
  color: #24292f;
  margin-bottom: 4px;
}

<span
  class="c0"
  color="fg.default"
>
  Hello
</span>
`;
